src/expressions/functions/builtInFunctions_fontoxpath.ts

```diff
@@ -1,23 +1,140 @@
+import { ElementNodePointer } from '../../domClone/Pointer';
 import { printAndRethrowError } from '../../evaluationUtils/printAndRethrowError';
-import astHelper from '../../parsing/astHelper';
+import astHelper, { IAST } from '../../parsing/astHelper';
 import compileAstToExpression from '../../parsing/compileAstToExpression';
+import convertXmlToAst from '../../parsing/convertXmlToAst';
 import parseExpression from '../../parsing/parseExpression';
 import processProlog from '../../parsing/processProlog';
 import annotateAst from '../../typeInference/annotateAST';
 import { AnnotationContext } from '../../typeInference/AnnotationContext';
 import createAtomicValue from '../dataTypes/createAtomicValue';
+import ISequence from '../dataTypes/ISequence';
+import isSubtypeOf from '../dataTypes/isSubtypeOf';
 import MapValue from '../dataTypes/MapValue';
 import sequenceFactory from '../dataTypes/sequenceFactory';
-import Value, { SequenceMultiplicity, ValueType } from '../dataTypes/Value';
+import Value, { SequenceMultiplicity, ValueType, valueTypeToString } from '../dataTypes/Value';
 import DynamicContext from '../DynamicContext';
+import ExecutionParameters from '../ExecutionParameters';
 import ExecutionSpecificStaticContext from '../ExecutionSpecificStaticContext';
 import { BUILT_IN_NAMESPACE_URIS } from '../staticallyKnownNamespaces';
 import StaticContext from '../StaticContext';
 import createDoublyIterableSequence from '../util/createDoublyIterableSequence';
 import { IIterator, IterationHint } from '../util/iterators';
+import { errXPTY0004 } from '../XPathErrors';
 import { BuiltinDeclarationType } from './builtInFunctions';
 import FunctionDefinitionType from './FunctionDefinitionType';
 
+function createAstFromValue(queryValue: Value, debug: boolean): IAST {
+	if (isSubtypeOf(queryValue.type, ValueType.XSSTRING)) {
+		return parseExpression(queryValue.value as string, {
+			allowXQuery: false,
+			debug,
+		});
+	}
+
+	if (isSubtypeOf(queryValue.type, ValueType.ELEMENT)) {
+		const nodePointer: ElementNodePointer = queryValue.value;
+
+		try {
+			return convertXmlToAst(nodePointer.node);
```

Comment:
Should this be able to handle a mix of nodes constructed in the query itself and (clones of) nodes existing in the context? That would enable modifying XQuery using XQuery, which is kind of neat. It seems that by taking the node out of the pointer we're discarding such information. I'm fine with leaving that out for now, but I do think in that case we should throw if the nodePointer is a structure that has any grafts.

Reply:
True, but lets worry about that when we get there.